src/gtest/test_checkblock.cpp

```diff
@@ -67,27 +67,99 @@ TEST(CheckBlock, BlockSproutRejectsBadVersion) {
+        // Revert to test default. No-op on mainnet params.
+        UpdateNetworkUpgradeParameters(Consensus::UPGRADE_OVERWINTER, Consensus::NetworkUpgrade::NO_ACTIVATION_HEIGHT);
+        UpdateNetworkUpgradeParameters(Consensus::UPGRADE_SAPLING, Consensus::NetworkUpgrade::NO_ACTIVATION_HEIGHT);
+    }
+
+    // Returns a valid but empty mutable transaction at block height 1.
+    CMutableTransaction GetFirstBlockTransaction() {
+        CMutableTransaction mtx;
+
+        // No inputs.
+        mtx.vin.resize(1);
+        mtx.vin[0].prevout.SetNull();
+
+        // Set height to 1.
+        mtx.vin[0].scriptSig = CScript() << 1 << OP_0;
+
+        // Give it a single zero-valued, always-valid output.
+        mtx.vout.resize(1);
+        mtx.vout[0].scriptPubKey = CScript() << OP_TRUE;
+        mtx.vout[0].nValue = 0;
+
+        // Give it a Founder's Reward vout for height 1.
+        mtx.vout.push_back(CTxOut(
+                    GetBlockSubsidy(1, Params().GetConsensus())/5,
+                    Params().GetFoundersRewardScriptAtHeight(1)));
+
+        return mtx;
+    }
+
+    // Expects a height-1 block containing a given transaction to pass
+    // ContextualCheckBlock. This is used in accepting (Sprout-Sprout,
+    // Overwinter-Overwinter, ...) tests. You should not call it without
+    // calling a SCOPED_TRACE macro first to usefully label any failures.
+    void ExpectValidBlockFromTx(const CTransaction& tx) {
+        // Create a block and add the transaction to it.
+        CBlock block;
+        block.vtx.push_back(tx);
+
+        // Set the previous block index to the genesis block.
+        CBlockIndex indexPrev {Params().GenesisBlock()};
+
+        // We now expect this to be a valid block.
+        MockCValidationState state;
+        EXPECT_TRUE(ContextualCheckBlock(block, state, &indexPrev));
+    }
+
+    // Expects a height-1 block containing a given transaction to fail
+    // ContextualCheckBlock. This is used in rejecting (Sprout-Overwinter,
+    // Overwinter-Sprout, ...) tests. You should not call it without
+    // calling a SCOPED_TRACE macro first to usefully label any failures.
+    void ExpectInvalidBlockFromTx(const CTransaction& tx, std::string reason) {
+        // Create a block and add the transaction to it.
+        CBlock block;
+        block.vtx.push_back(tx);
+
+        // Set the previous block index to the genesis block.
+        CBlockIndex indexPrev {Params().GenesisBlock()};
+
+        // We now expect this to be an invalid block, for the given reason.
+        MockCValidationState state;
+        EXPECT_CALL(state, DoS(100, false, REJECT_INVALID, reason, false)).Times(1);
+        EXPECT_FALSE(ContextualCheckBlock(block, state, &indexPrev));
+    }
+
+};
+
+
+TEST_F(ContextualCheckBlockTest, BadCoinbaseHeight) {
+    // Put a zero-height transaction in a block
+    CMutableTransaction mtx = GetFirstBlockTransaction();
+    mtx.vin[0].scriptSig = CScript() << 0 << OP_0;
```

Comment:
This is altered from the original test, which does not put a zero height in the `scriptSig` (matching our consensus rule, which is not enforced on genesis blocks).